android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java

```diff
@@ -692,23 +721,47 @@ private String getSubkeyText(String keyID, String keyText) {
     return text;
   }
 
-  private String makeSvgOnlyFont(String font) {
+  /**
+   * Take a font JSON object and adjust to pass to JS
+   * 1. Replace "source" keys for "files" keys
+   * 2. Create full font paths for .ttf or
+   * .svg for Android 4.2 or 4.3 OSK
+   * @param font String font JSON object as a string
+   * @return String of modified font information with full paths. If font is invalid, return "undefined"
+   */
+  private String makeFontPaths(String font) {
     try {
       JSONObject fontObj = new JSONObject(font);
-      JSONArray sourceArray = fontObj.optJSONArray(KMManager.KMKey_FontFiles);
-      if (sourceArray != null) {
-        String fontFile;
-        int length = sourceArray.length();
-        for (int i = 0; i < length; i++) {
-          fontFile = sourceArray.getString(i);
-          if (fontFile.contains(".svg")) {
-            fontObj.put(KMManager.KMKey_FontFiles, fontFile);
-            return fontObj.toString();
+      JSONArray sourceArray;
+      String fontFile;
+
+      // Replace "sources" key with "files"
+      if (fontObj.has(KMManager.KMKey_FontSource)) {
+        fontObj.put(KMManager.KMKey_FontFiles, fontObj.get(KMManager.KMKey_FontSource));
+        fontObj.remove(KMManager.KMKey_FontSource);
+      }
+
+      Object obj = fontObj.get(KMManager.KMKey_FontFiles);
+      if (obj instanceof String) {
+        fontFile = fontObj.getString(KMManager.KMKey_FontFiles);
+        fontObj.put(KMManager.KMKey_FontFiles, keyboardRoot + obj);
+        return fontObj.toString();
+      } else if (obj instanceof JSONArray) {
+        sourceArray = fontObj.optJSONArray(KMManager.KMKey_FontFiles);
+        if (sourceArray != null) {
+          for (int i = 0; i < sourceArray.length(); i++) {
+            fontFile = sourceArray.getString(i);
+            if (fontFile.contains(".ttf") ||
+              (fontFile.contains(".svg") && (Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN_MR1 || Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN_MR2))) {
+              fontObj.put(KMManager.KMKey_FontFiles, keyboardRoot + fontFile);
+              fontObj.remove(KMManager.KMKey_FontSource);
+              return fontObj.toString();
+            }
           }
         }
       }
     } catch (JSONException e) {
-      return "''";
+      return "undefined";
```

Comment:
why the change to `undefined`?

Reply:
Invalid font strings are passed as `"undefined"` in the JS calls to KMW